src/domain/Processors/VehicleProcessor.ts

```diff
@@ -0,0 +1,533 @@
+      this.techRecordsListHandler
+    );
+  }
+
+  /* #region  Protected functions */
+  /**
+   * Validates tech record based on vehicle type and returns string array of error messages
+   * @param techRecord the techRecord to be validated
+   */
+  protected abstract validateTechRecordFields(techRecord: TechRecord, isCreate?: boolean): string[];
+  /**
+   * map attributes to the tech record based on vehicle type
+   * @param techRecord techRecord to be mapped
+   */
+  protected abstract mapFields(techRecord: TechRecord): TechRecord;
+
+  /**
+   * Returns the system number or trailer id based on vehicle type
+   */
+  protected abstract async setNumberKey(): Promise<string>;
+
+  /**
+   * Calculate record completeness based on vehicle type.
+   */
+  protected calculateRecordCompleteness(vehicle: T): string {
+    return computeRecordCompleteness(vehicle);
+  }
+
+  /**
+   * update the vehicle identifier attributes
+   * @param existingVehicle The existing tech record
+   * @param updatedVehicle The updated tech record
+   */
+  protected updateVehicleIdentifiers(updatedVehicle: T, existingVehicle: T) {
+    const { primaryVrm, secondaryVrms } = updatedVehicle;
+    if (secondaryVrms) {
+      existingVehicle.secondaryVrms = secondaryVrms;
+    }
+    if (primaryVrm && existingVehicle.primaryVrm !== primaryVrm) {
+      const previousVrm = existingVehicle.primaryVrm;
+      if (previousVrm) {
+        existingVehicle.secondaryVrms = existingVehicle.secondaryVrms
+          ? existingVehicle.secondaryVrms
+          : [];
+        existingVehicle.secondaryVrms.push(previousVrm);
+      }
+      existingVehicle.primaryVrm = primaryVrm;
+      updatedVehicle.techRecord[0].reasonForCreation =
+        `VRM updated from ${previousVrm} to ${primaryVrm}. ` +
+        updatedVehicle.techRecord[0].reasonForCreation;
+    }
+  }
+
+  /**
+   * Validate VRMs in payload with existing VRMs
+   * @param newVehicle TechRecord received in a payload
+   * @param existingVehicle TechRecord in the database
+   */
+  protected async validateVrmWithHistory(newVehicle: T, existingVehicle: T) {
+    const errors: string[] = [];
+    const { primaryVrm } = newVehicle;
+    if (primaryVrm && existingVehicle.primaryVrm !== primaryVrm) {
+      const primaryVrmRecords = await this.techRecordDAO.getBySearchTerm(
+        primaryVrm,
+        SEARCHCRITERIA.VRM
+      );
+      if (primaryVrmRecords.Count! > 0) {
+        errors.push(`Primary VRM ${primaryVrm} already exists`);
+      }
+    }
+    return errors;
+  }
+
+  /**
+   * Set all vehicle identifiers to upper case
+   * @param vehicle the input parameter
+   */
+  protected capitaliseGeneralVehicleAttributes(vehicle: T) {
+    vehicle.vin = vehicle.vin?.toUpperCase();
+    vehicle.partialVin = vehicle.partialVin?.toUpperCase();
+    vehicle.primaryVrm = vehicle.primaryVrm?.toUpperCase();
+    vehicle.secondaryVrms = vehicle.secondaryVrms?.map((vrm: string) =>
```

Comment:
I leave it you but we could create an util function capitalize since we are reusing a lot like
const toCapitalize = (st) => str.toUpperCase();
and the secondaryVrms (and rest) we apply it `.map(toCapitalize) `

Reply:
updated `toUpperCase` with `toCapitalise`

Reply:
reused `toUpperCase` as the map function name as toCapitalise has different behavior